src/decoder/plugins/UsfDecoderPlugin.cxx

```diff
@@ -0,0 +1,318 @@
+    }
+    return 0;
+}
+
+
+inline static double
+track_lengths(double length, double fade)
+{
+    return length + fade;
+
+}
+
+static void
+usf_file_decode(Decoder &decoder, Path path_fs)
+{
+    /* Load the file */
+
+    UsfTags tags;
+    UsfLoaderState state;
+    state.emu = malloc(usf_get_state_size());
+    usf_clear(state.emu);
+    AtScopeExit(state) {
+        free(state.emu);
+    };
+
+    const char* path = path_fs.c_str();
+    const int psf_version_tags = psf_load(path, &stdio_callbacks, 0, 0, 0, usf_tags_target, &tags, 0);
+    const int psf_version_state = psf_load( path, &stdio_callbacks, psf_version_tags, usf_loader, &state, usf_info, &state, 0 );
+
+    if ( psf_version_tags != 0x21 || psf_version_state <= 0 ) {
+        LogWarning(usf_domain, "Error loading usf file");
+        return;
+    }
+
+    usf_set_compare( state.emu, state.enableCompare );
+    usf_set_fifo_full( state.emu, state.enableFIFOFull );
+
+    int32_t sample_rate;
+    usf_render( state.emu, 0, 0, &sample_rate );
+
+    /* initialize the MPD decoder */
+
+    const AudioFormat audio_format(sample_rate, SampleFormat::S16, USF_CHANNELS);
+    assert(audio_format.IsValid());
+
+    double totalLength =  track_lengths(tags.length, tags.fade);
+
+    // Duration
+    decoder_initialized(decoder, audio_format, true, SongTime::FromS(totalLength));
+
+    /* .. and play */
+    DecoderCommand cmd;
+    do {
+        int16_t buf[USF_BUFFER_SAMPLES];
+        const char* result = usf_render(state.emu, buf, USF_BUFFER_FRAMES, 0);
+        if (result != 0) {
+            LogWarning(usf_domain, result);
+            break;
+        }
+
+        // Simple fading
+        double fadeTime = tags.fade;
+        double trackLength = tags.length;
+        if (fadeTime > 0 && totalLength >= 0 && decoder.timestamp > trackLength) {
+            const double vol = 1.0 - ((decoder.timestamp - trackLength) / fadeTime);
+            const double normalizedVol = vol < 0 ? 0 : vol;
+            for (unsigned int i = 0; i < USF_BUFFER_SAMPLES; i++) {
+                buf[i] *= normalizedVol;
+            }
+        }
+
+        cmd = decoder_data(decoder, nullptr, buf, sizeof(buf), 0);
+
+        // Stop song manually
+        if (totalLength >= 0 && decoder.timestamp > totalLength+2)
+            break;
+
+        if (cmd == DecoderCommand::SEEK) {
+            // If user seeks during the fade period. Disable fading and play forever.
+            // Hacky way to give user posibility to enable looping on the fly
+            if (decoder.timestamp > trackLength) {
+                totalLength = -1;
+            }
+            // Seek manually by restarting emulator and discarding samples.
+            const double target_time = decoder_seek_time(decoder).ToDoubleS();
+            usf_restart(state.emu);
+            const unsigned frames_to_throw = (unsigned) (sample_rate*target_time+0.5);
+            usf_render(state.emu, 0, frames_to_throw, 0);
```

Comment:
Does this mean that by seeking to the end of a large file, the decoder will decode the whole file completely, only to go to that position? I'd say this decoder is not seekable.